orttraining/orttraining/python/training/optim/lr_scheduler.py

```diff
@@ -0,0 +1,290 @@
+
+        Returns:
+            ordered :py:obj:`list` of learning rates.
+                The first entry is the default learning rate and
+                    the remaining, refer to each :py:class:`optim._OptimizerConfig`s parameter group.
+            NOTE: Currently, only the default learning rate is supported,
+                which implies returning a list with a single value.
+        """
+        raise NotImplementedError
+
+    def get_last_lr(self):
+        r""" Return last computed learning rate by LR Scheduler"""
+        return self._last_lr
+
+    def _step(self, train_step_info):
+        r"""Private method called to update :py:class:`optim._OptimizerConfig` learning rate
+
+        NOTE: This class should never be called by the user.
+        """
+        new_lr = self.get_lr(train_step_info)
+        self._last_lr = new_lr
+        self.optimizer_config.lr = new_lr[0]
+
+
+class ConstantWarmupLRScheduler(LRScheduler):
+    r"""Constant warmup strategy for learning rate update
+
+    Learning rate update strategy:
+        lr = base_lr * (step / total_steps) / warmup, when step / total_steps < warmup
+        lr = base_lr, when step / total_steps >= warmup
+
+    Args:
+        optimizer_config (optim._OptimizerConfig): optimizer config.
+            One of :py:class:`.optim.Adam`, :py:class:`.optim.Lamb` or :py:class:`.optim.SGD`.
+        total_steps (int): total training steps for learning.
+        warmup (float, default is 0.002): portion of total steps for warmup. Range is (0, 1]
+
+    Example:
+        .. code-block:: python
+
+            # Initialize optimizer config
+            optimizer_config = optim.SGD(lr=0.001)
+
+            # Initialize lr scheduler
+            lr_scheduler = ConstantWarmupLRScheduler(optimizer_config, total_steps=512, warmup=0.002)
+
+            # Initialize ORTTrainer with lr scheduler
+            opts = ORTTrainerOptions({
+                lr_scheduler: lr_scheduler
+            })
+            ort_trainer = ORTTrainer(..., options=opts)
+
+            # Call step() in every batch update
+            for inputs in batch_inputs:
+                outputs = ort_trainer.train_step(**inputs)
+    """
+
+    def __init__(self, optimizer_config, total_steps, warmup=0.002):
+        super().__init__(optimizer_config)
+        assert isinstance(total_steps, int) and total_steps > 0,\
+            "total_steps must be a strict positive number"
+        assert isinstance(warmup, float) and warmup >= 0 and warmup < 1,\
+            "warmup must be a float between (0, 1]"
+        assert total_steps > warmup,\
+            "total_steps must be greater than warmup"
+
+        self.total_steps = total_steps
+        self.warmup = warmup
+
+    def _warmup_constant(self, train_step_info):
+        # Adds 1 to train_step_info.step and self.total_steps to prevent zero'ing lr
+        x = (train_step_info.step + 1) / (self.total_steps + 1)
```

Comment:
How does this warmup strategy (especially the +1 in num and den) compare with PyTorch?

Reply:
Pytorch has a different approach and don't use `step` we do, probably because their scheduler don't use the warmup concept.

In their approach, they rely on `epoch` instead. When `epoch > 0`, compute is done using `epoch` as part of the input, otherwise, the initial lr is returned without any compute (reference https://github.com/pytorch/pytorch/blob/master/torch/optim/lr_scheduler.py#L310)

Reply:
Are we able to create lr schedulers that match PyTorch behavior?

Reply:
Yes, it should as both `epoch` and `step` are available at `LRSCheduler.update(self, train_step_info)` and other scheduler specific parameters can be passed in to the LRScheduler constructor.